pkg/kv/kvserver/replica_application_result.go

```diff
@@ -366,17 +343,30 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		raftAdmissionMeta:       nil,
 		v2SeenDuringApplication: false,
 	}
-	// If the original proposal had an explicit span, it's an async consensus
-	// proposal and the span would be finished momentarily (when we return to
-	// the caller) if we didn't unlink it here, but we want it to continue
-	// tracking newProposal. We leave it in `origP.ctx` though, since that
-	// context will become unused once the application of this (soft-failed)
-	// proposal concludes, i.e. soon after this method returns, in case there
-	// is anything left to log into it.
+
 	defer func() {
-		if success {
-			origP.sp = nil
+		if !success {
+			return
 		}
+		// If the original proposal had an explicit span, it's an async consensus
+		// proposal and the span would be finished momentarily (when we return to
+		// the caller) if we didn't unlink it here, but we want it to continue
+		// tracking newProposal. We leave it in `origP.ctx` though, since that
+		// context will become unused once the application of this (soft-failed)
+		// proposal concludes, i.e. soon after this method returns, in case there is
+		// anything left to log into it.
+		origP.sp = nil
+		origP.quotaAlloc = nil
+		origP.ec = makeEmptyEndCmds()
+		origP.doneCh = nil
```

Comment:
Alternatively, we could put all these (maybe a few other) fields, and the context, to a struct, and reference the whole struct from `ProposalData` instead of moving/copying things between proposals.

Reply:
I think it's fine to keep this as it is.